src/utils/wallet.js

```diff
@@ -82,19 +84,77 @@ export function signTransaction(trx: Object, walletInstance: Object): Promise<st
   return wallet.signTransaction(signTx);
 }
 
+export function encodePersonalMessage(message: string): string {
+  const data = toBuffer(convertUtf8ToHex(message));
+
+  const buf = Buffer.concat([
+    Buffer.from("\x19Ethereum Signed Message:\n" + data.length.toString(), "utf8"), // eslint-disable-line
+    data,
+  ]);
+
+  return bufferToHex(buf);
+}
+
+export function hashPersonalMessage(message: string): string {
+  const data = encodePersonalMessage(message);
+
+  const buf = toBuffer(data);
+  const hash = keccak256(buf);
+
+  return bufferToHex(hash);
+}
+
 // handle eth_sign
 export function signMessage(message: any, walletInstance: Object): string {
   const provider = getEthereumProvider(getEnv().NETWORK_PROVIDER);
   const wallet = walletInstance.connect(provider);
-  // TODO: this method needs to be replaced when ethers.js is migrated to v4.0
-  return ethSign(message, wallet.privateKey);
+
+  const data = isHexString(message) ? ethers.utils.arrayify(message) : message;
+
+  return wallet.signMessage(data);
 }
 
 // handle personal_sign
-export function signPersonalMessage(message: string, walletInstance: Object): Promise<string> {
+export function signPersonalMessage(messageHex: string, walletInstance: Object): Promise<string> {
+  const provider = getEthereumProvider(getEnv().NETWORK_PROVIDER);
+  const wallet = walletInstance.connect(provider);
+
+  const actualMessage = ethers.utils.toUtf8String(messageHex);
+  const data = hashPersonalMessage(actualMessage);
+
+  return wallet.signMessage(ethers.utils.arrayify(data));
+}
+
+export function encodeTypedDataMessage(message: string): string {
+  const useV4 = true;
+  const data = TypedDataUtils.sanitizeData(JSON.parse(message));
```

Comment:
Just wondering if `message` in `JSON.parse` is always parsable? JSON.parse fails and throws spectacularly when it cannot parse. Might be worth wrapping in a try / catch?

Reply:
Yes, it's always ABI, otherwise it's failure on dapp end, but we can wrap for sure 👍

Reply:
Actually, it's a chain of sign calls and the top method call (2 of them for separate chain of calls) that's waiting for result has exception wrapper so we're fine.

https://github.com/pillarwallet/pillarwallet/blob/99cf16c30426394fefb478af8348e3280b3de127/src/screens/WalletConnect/WalletConnectPinConfirm.js#L163-L202

Reply:
Thanks for the clarification! 👍